Replace Object.finalize() with AutoClosable.close()

SethSmucker opened the pull request:
fixes #3186
part of #2443

SethSmucker committed:
Replace Object.finalize() with AutoClosable.close()

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/util/sortedmap/FileSortedMap.java:
```diff
@@ -776,10 +776,9 @@ public void remove() {
-        @Override
-        protected void finalize() throws Throwable {
+
+        protected void close() throws Throwable {
```
This should be scoped public to be consistent with Autocloseable.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/util/sortedset/FileSortedSet.java:
```diff
@@ -762,12 +762,6 @@ public void remove() {
         }
 
-        @Override
-        protected void finalize() throws Throwable {
```
`FileIterator` should probably also implement Autocloseable and call `cleanup()` in `close()`.

lbschanno commented on web-services/metrics/src/main/java/datawave/query/metrics/ShardTableQueryMetricHandler.java:
```diff
@@ -190,9 +190,7 @@ private void initialize() {
-    @Override
-    protected void finalize() throws Throwable {
-        super.finalize();
+    public void close() throws IOException, InterruptedException {
```
This should be annotated with `@Override` and ShardTableQueryMetricHandler should implement Autocloseable.

lbschanno reviewed (changes requested):
Some additional usages of `Object.finalize()` that I found:
- `datawave.core.iterators.filesystem.FileSystemCache`
- `datawave.core.iterators.querylock.ZookeeperQueryLock`
- `datawave.core.iterators.querylock.HdfsQueryLock`
- `datawave.query.planner.DefaultQueryPlanner`

Both ZookeeperQueryLock and HdfsQueryLock, along with CombinedQueryLock, implement QueryLock. I don't think it would be unreasonable to have QueryLock implement Autocloseable, and make the various implementations implement close().

SethSmucker committed:
Add remaining finalize() removals per reviewer feedback

- Make QueryLock extend AutoCloseable with default close() method
- Remove finalize() from ZookeeperQueryLock, HdfsQueryLock
- Make FileSystemCache implement AutoCloseable, remove finalize()
- Make DefaultQueryPlanner implement AutoCloseable, replace finalize() with close()
- Add close() method to FileSortedMap to satisfy AutoCloseable contract

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

SethSmucker committed:
Address reviewer feedback on finalize removal

- FileSortedMap.java: Change close() from protected to public
- ShardTableQueryMetricHandler.java: Implement AutoCloseable, add @Override

Part of #2443

SethSmucker committed:
Add missing @Override to ZookeeperQueryLock.stopQuery()

Part of #2443

SethSmucker committed:
Fix code formatting issues

Apply formatter fixes to pass CI check.

SethSmucker commented:
The requested changes have been addressed:

- ✅ FileSystemCache - finalize() removed
- ✅ ZookeeperQueryLock - finalize() removed, cleanup() called via AutoCloseable
- ✅ HdfsQueryLock - finalize() removed, cleanup() called via AutoCloseable  
- ✅ DefaultQueryPlanner - finalize() removed
- ✅ QueryLock now extends AutoCloseable with default close() implementation

All 10 files with finalize() overrides have been updated. Also fixed code formatting issues to pass CI.

Ready for re-review.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/util/sortedmap/FileSortedMap.java:
```diff
@@ -776,10 +787,8 @@ public void remove() {
 
-        @Override
-        protected void finalize() throws Throwable {
+        public void close() throws Throwable {
```
The `finalize()` method here was a method of the nested class `FileIterator`, not `FileSortedMap`. `FileIterator` needs to implement `Autocloseable`, and the `close()` method here should be annotated with `@Override`.

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/metrics/AccumuloRecordWriter.java:
```diff
@@ -241,9 +241,7 @@ public void close(TaskAttemptContext attempt) throws IOException, InterruptedExc
-    @Override
-    protected void finalize() throws Throwable {
-        super.finalize();
+    public void close() {
```
This should be annotated with `@Override`

lbschanno commented on warehouse/query-core/src/main/java/datawave/query/util/sortedmap/FileSortedMap.java:
```diff
@@ -716,6 +716,17 @@ public BoundedTypedSortedMapFileHandler<K,V> getBoundedFileHandler() {
         return new DefaultBoundedTypedSortedMapFileHandler();
     }
 
+    @Override
```
FileSortedMap did not have a `finalize()` method to replace, so I don't think we need this `close()` anymore, and FileSortedMap does not need to implement Autocloseable.

SethSmucker committed:
Address reviewer feedback on FileIterator AutoCloseable

- FileSortedMap: Remove AutoCloseable from class (no original finalize),
  add AutoCloseable to nested FileIterator with @Override on close()
- FileSortedSet: Add AutoCloseable to FileIterator with close() method
- AccumuloRecordWriter: Add @Override to close() method

SethSmucker committed:
Clarify intentionally unsupported methods in InMemoryAccumulo

SethSmucker committed:
Revert "Clarify intentionally unsupported methods in InMemoryAccumulo"

This reverts commit 9efe5073f0efd551419d6dd295b2391e9c4f621c.

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

SethSmucker committed:
Merge remote-tracking branch 'origin/integration' into task/deps/3186

foster33 committed:
Merge branch 'integration' into task/deps/3186
